templates/minimal/runtime/tests/validate_dev_chain_spec.rs

```diff
@@ -0,0 +1,59 @@
+use polkadot_sdk::*;
+use minimal_template_runtime::WASM_BINARY;
+use sc_service::ChainType;
+use sc_chain_spec::{
+    ChainSpecExtension,
+    ChainSpecGroup,
+    Properties,
+};
+use serde::{Deserialize, Serialize};
+use core::str;
+use std::io::BufReader;
+
+pub type ChainSpec = sc_service::GenericChainSpec<Extensions>;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ChainSpecGroup, ChainSpecExtension)]
+pub struct Extensions {
+    #[serde(alias = "relayChain", alias = "RelayChain")]
+    pub relay_chain: String,
+    #[serde(alias = "paraId", alias = "ParaId")]
+    pub para_id: u32,
+}
+
+impl Extensions {
+    pub fn try_get(chain_spec: &dyn sc_service::ChainSpec) -> Option<&Self> {
+        sc_chain_spec::get_extension(chain_spec.extensions())
+    }
+}
+
+#[test]
+fn test_minimal_dev_chain_spec_rt_validity() {
+    let mut properties = Properties::new();
+    properties.insert("tokenDecimals".to_string(), 12.into());
+    properties.insert("tokenSymbol".to_string(), "UNIT".into());
+    let current_wasm = WASM_BINARY.expect("Development wasm not available");
+
+    let test_chain_spec: serde_json::Value = serde_json::from_str(&ChainSpec::builder(
+        current_wasm,
+        Extensions {
+            relay_chain: "dev".into(),
+            para_id: 1000,
+        })
+        .with_name("Custom")
+        .with_id("custom")
+        .with_chain_type(ChainType::Live)
+        .with_genesis_config_preset_name(sp_genesis_builder::DEV_RUNTIME_PRESET)
+        .with_properties(properties.clone())
+        .build()
+        .as_json(false)
+        .unwrap()).unwrap();
+
+    let existing_chain_spec_file =
+		std::fs::File::open("../dev_chain_spec.json").expect("file should open. qed");
+	let existing_chain_spec_reader = BufReader::new(existing_chain_spec_file);
+	let existing_chain_spec: serde_json::Value =
+		serde_json::from_reader(existing_chain_spec_reader).expect("should read proper JSON. qed");
+
+    assert_eq!(existing_chain_spec, test_chain_spec);
```

Comment:
As I was afraid, running the test on my machine failed. Confirmed with a few more people that building the runtimes isn't deterministic yet. We could use `srtool`: https://github.com/chevdor/srtool to get deterministic runtimes and as a result deterministic `dev_chain_spec.json`s. There are two approaches from what I can tell:

1. Use `srtool-cli` to generate the `dev_chain_spec.json` locally, and when it comes to the test that should run in the CI, we'd rely on a github action that installs srtool-cli, runs it against the runtimes to build the wasm blob, and then installs `chain-spec-builder` and uses it against the built `.wasm` to generate the `dev_chain_spec.json`. Then finally runs a diff between the CI `dev_chain_spec.json` and the one from the templates.

2. Use `srtool-action` as we use it here: https://github.com/paritytech/polkadot-sdk/blob/master/.github/workflows/release-30_publish_release_draft.yml#L53. IMO it is not doing exactly what we want since it favors cases where we build the majority of the existing runtimes (because of the `excluded runtimes` logic, which is easier to use when the excluded runtimes list is small, or known), instead of building only certain known runtimes. IMO it needs some adjustments to make it work easily for our use case too, and then once the runtimes are built (and corresponding wasm blobs are uploaded as artifacts), download the artifacts, install `chain-spec-builder`, generate the `dev_chain_spec.json` and compare it with the existing `dev_chain_spec.json`, as in the first case.

Neither of the options is straightforward, and as you can notice, we'd change the direction from a Rust test to a github workflow. Doing either of the options could be slow too given the GH actions that run on this PR need to be approved by a maintainer first, and that impacts your feedback cycle loop. I know it might be disappointing coming with such a big request for a change at this late hour, and I get it if you'd like to stop here for now, but if you can still invest time in this LMK, I would be happy to assist.

Reply:
Hey @iulianbarbu, I will implement the strategies you suggested. I will reach out if I have any questions.

Reply:
Writing the integration test, was a dope opportunity to learn the product⚡️.  How the different components interact.  I enjoyed it. Great onboarding task